### PR TITLE
Promoted File#path to IO#path, added path: optional param to IO#new

### PR DESCRIPTION
**Pull request by moste00**
Implementing feature [Feature #19036](https://bugs.ruby-lang.org/issues/19036)] in Ruby 3.2 compatibility.

### CODE REVIEW

File: src/main/ruby/truffleruby/core/truffle/io_operations.rb [hunk 1]
```diff
@@ -543,10 +543,14 @@ def self.normalize_options(mode, perm, options, default_mode = nil)
             external, internal = encoding.split(':', 2)
           end
         end
+
+        if options[:path]
+          path = options[:path]
+        end
```
Comment (andrykonchin) [inline src/main/ruby/truffleruby/core/truffle/io_operations.rb:549]: CRuby converts `path` to String with `#to_str` (and raises error if cannot) and freezes result (but it seems freezing cannot be observer outside of an IO object and it's kind of optimization):

```c
    if (!NIL_P(opt)) {
        if (rb_hash_aref(opt, sym_autoclose) == Qfalse) {
            fmode |= FMODE_PREP;
        }

        path = rb_hash_aref(opt, RB_ID2SYM(idPath));
        if (!NIL_P(path)) {
            StringValue(path);
            path = rb_str_new_frozen(path);
        }
    }
```

Converting to String could be done with helpers like `Truffle::Type.rb_convert_type`

Comment (andrykonchin) [reply src/main/ruby/truffleruby/core/truffle/io_operations.rb:549]: Also `if options[:path]` isn't precise enough - it treats `false` similar to `nil` and keeps the `path` variable unassigned. But CRuby will check `nil` value (with `if (!NIL_P(path)) {`) and try to convert `false` to String anyway.

`Primitive.nil?` could be used instead of `#nil?` Ruby method.

File: src/main/ruby/truffleruby/core/io.rb [hunk 2]
```diff
@@ -857,6 +857,7 @@ def initialize(fd, mode = nil, **options)
 
     @autoclose = autoclose_tmp
     @pipe = false
+    @path = path unless @path
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/io.rb:860]: This feels hacky, there should be no ` unless @path` here.
You should pass the path there to as a keyword argument to `super`.

**Review by eregon — COMMENTED** [review]
Thanks for the PR.
You need to untag the spec for this with `jt untag spec/ruby/core/io/path_spec.rb`.

File: src/main/ruby/truffleruby/core/io.rb (hunk 2 above)
Comment (andrykonchin) [inline src/main/ruby/truffleruby/core/io.rb:860]: **suggestion:** Here we ignore `path` option if path is already set. I would prefer to keep this logic (I mean "ignoring") in ancestor classes.

For instance for `File` we could just reorder `super` call and assigning the `@path` variable:

```ruby
  def initialize(path_or_fd, mode = nil, perm = nil, **options)
    # ...
      super(fd, mode, **options)
      @path = path
    # ...
  end
```

**Comment by andrykonchin** [comment]
It makes sense to declare in this PR a `#to_path` method, that is just an alias to `#path`.

File: src/main/ruby/truffleruby/core/io.rb (hunk 2 above)
Comment (moste00) [reply src/main/ruby/truffleruby/core/io.rb:860]: Could you elaborate on that a bit ? Do you mean that IO should still keep the "ignoring path unless set" and the descendent class File should then just set it itself ? 

I rather prefer @eregon suggestion, it's more elegant for File to just use the API of its parent like everybody else by overriding path in the **options dictionary, leaving the assignment to be done by IO.

File: src/main/ruby/truffleruby/core/truffle/io_operations.rb (hunk 1 above)
Comment (moste00) [reply src/main/ruby/truffleruby/core/truffle/io_operations.rb:549]: Ah, good catch. Thanks! 

(Although one wonders who on earth would prefer "false" to be converted to a file path. But API promises are API promises and whatever MRI does is the spec I guess.)

**Comment by moste00** [comment]
> It makes sense to declare in this PR a #to_path method as well, that is just an alias of #path.

You mean in IO ? File already declares a to_path using alias_method

**Comment by andrykonchin** [comment]
Yes, I mean in `IO`. CRuby defines `#to_path` in the `IO` class:

```ruby
f = File.new("a.txt", "w")

f.method(:path).owner
#=> IO
f.method(:to_path).owner
#=> IO
```

File: src/main/ruby/truffleruby/core/io.rb (hunk 2 above)
Comment (andrykonchin) [reply src/main/ruby/truffleruby/core/io.rb]: I also like @eregon's suggestion.

**Review by andrykonchin — APPROVED** [review]

**Comment by andrykonchin** [comment]
Could you please rename the commit, e.g. name it similar to the PR title?

File: src/main/ruby/truffleruby/core/file.rb [hunk 3]
```diff
@@ -37,9 +37,12 @@
...
   include Enumerable
 
-  class FileError < Exception; end # rubocop:disable Lint/InheritException
+  class FileError < RuntimeError; end
+
   class NoFileError < FileError; end
+
   class UnableToStat < FileError; end
+
   class PermissionError < FileError; end
```
Comment (andrykonchin) [inline src/main/ruby/truffleruby/core/file.rb:46]: It seems these exception classes aren't used anywhere. Should we just remove them?

cc @eregon

File: src/main/ruby/truffleruby/core/file.rb [hunk 4]
```diff
@@ -37,9 +37,12 @@
 class File < IO
   include Enumerable
 
-  class FileError < Exception; end # rubocop:disable Lint/InheritException
+  class FileError < RuntimeError; end
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/file.rb:40]: Why did you change this?
Such change has nothing to do with this PR AFAIK.
But as @andrykonchin says below, you could remove those 4 classes in a separate commit in this PR as a cleanup if you like, or undo these changes.

File: src/main/ruby/truffleruby/core/file.rb [hunk 5]
```diff
@@ -1161,13 +1164,12 @@ class << self
...
       super(path_or_fd, mode, **options)
-      @path = nil
     else
       path = Truffle::Type.coerce_to_path path_or_fd
       nmode, _binary, _external, _internal, _autoclose, perm = Truffle::IOOperations.normalize_options(mode, perm, options)
       fd = IO.sysopen(path, nmode, perm)
 
-      @path = path
+      options[:path] = path
       super(fd, mode, **options)
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/file.rb:1173]:
```suggestion
      super(fd, mode, path: path, **options)
```

File: src/main/ruby/truffleruby/core/io.rb [hunk 6]
```diff
@@ -1221,6 +1222,12 @@ def prepare_read_string(str)
     @lineno += 1
   end
 
+  def path
+    @path.dup
+  end
+
+  alias_method :to_path, :path
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/io.rb:1229]:
```suggestion
  end
  alias_method :to_path, :path
```
Let's keep them together that way it's easier to see what is aliased to what

File: src/main/ruby/truffleruby/core/file.rb (hunk 4 above)
Comment (moste00) [reply src/main/ruby/truffleruby/core/file.rb:40]: It was complained about by rubocop, and since rubocop was being run by the pre-commit hook, I couldn't commit without fixing it. I don't know why the hook let me commit before, and I don't remember pulling any new changes to the lint rules since I started the branch.

Anyway, them being not needed simplifies things a lot, I will delete them in a seperate commit and if the CI didn't fail then that's all good.

Comment (eregon) [reply src/main/ruby/truffleruby/core/file.rb:40]: Mmh, `jt rubocop` should pick the right version and config, and it does work for other people so I suspect something was broken in your dev env or so.

File: src/main/ruby/truffleruby/core/file.rb (hunk 5 above)
Comment (moste00) [reply src/main/ruby/truffleruby/core/file.rb:1173]: Hmm, I think this fails to override the passed value of :path if the user happens to pass it to ` File#new`. I tried a small experiment :

```
irb(main):001:1* def x(**options)
irb(main):002:1*   puts options
irb(main):003:0> end
=> :x
irb(main):004:1* def xoverride(**options)
irb(main):005:1*   x(overriden: "overriden", **options)
irb(main):006:0> end
=> :xoverride
irb(main):007:0> xoverride(option1: "option1", overriden: "no")
{:overriden=>"no", :option1=>"option1"}
=> nil      
```                                            
So it's indeed the passed argument that "wins". Whereas if you try : 

```
irb(main):014:0> File.new("/home/mostafa/test.rb", path: "invalid").path
=> "/home/mostafa/test.rb"
```
The desired behaviour is that the passed argument is ignored.

Comment (eregon) [reply src/main/ruby/truffleruby/core/file.rb:1173]: How about `super(fd, mode, **options, path: path)` then?

Comment (moste00) [reply src/main/ruby/truffleruby/core/file.rb:1173]: Works.

File: CHANGELOG.md [hunk 7]
```diff
@@ -13,6 +13,8 @@ Compatibility:
 * Add `Exception#detailed_message` method (#3257, @andrykonchin).
 * Fix `rb_enc_vsprintf` and force String encoding instead of converting it (@andrykonchin).
 * Add `rb_gc_mark_movable` function (@andrykonchin).
+* Promoted `File#path` and `File#to_path` to `IO#path` and `IO#to_path` and made IO#new accept an optional `path:` keyword argument (#3275, @moste00)
```
Comment (andrykonchin) [inline CHANGELOG.md:16]: It's convenient to use imperative verbs in a commit title, so it could be "Promote ...".

Also there may be mentioned a related GitHub issue (if it exists), not a link to a pull request.

**Comment by andrykonchin** [comment]
Could you please squash working commits? Removing exception classes might deserve a separate commit.

File: src/main/ruby/truffleruby/core/truffle/io_operations.rb [hunk 8]
```diff
@@ -543,10 +543,14 @@ def self.normalize_options(mode, perm, options, default_mode = nil)
             external, internal = encoding.split(':', 2)
           end
         end
+
+        unless Primitive.nil? options[:path]
+          path = StringValue(options[:path])
```
Comment (eregon) [inline src/main/ruby/truffleruby/core/truffle/io_operations.rb:548]: `path = options[:path]` above this `unless` would avoid two hash lookups in case it is set

File: CHANGELOG.md (hunk 7 above)
Comment (moste00) [reply CHANGELOG.md]: Hey, I know I totally left this PR for ages, sorry !

This is done now, I squashed all commits and edited both the CHANGELOG and the commit message to use "promote". 

Seperating the deletion of the unused exception classes into a seperate commit is a bit hairy/tedious git manipulation, so I figured I can get lazy and squash everything. If it's absolutely necessary, let me know and I will do it at once.

Comment (andrykonchin) [reply CHANGELOG.md]: Looks great!

Could you please address two comments that are left? (changelog record wording and `options[:path]` hash lookup)

Comment (moste00) [reply CHANGELOG.md]: Done now ! 

I have no idea how I tricked my brain into thinking that I already made the changes when I wrote the earlier comments haha.

Comment (andrykonchin) [reply CHANGELOG.md]: 👍

**Review by andrykonchin — APPROVED** [review]

**Review by eregon — APPROVED** [review]
